CMakeLists.txt

```diff
@@ -1,65 +1,138 @@
 cmake_minimum_required(VERSION 3.10)
 project(proc_image_processing)
 
-if (NOT CMAKE_BUILD_TYPE)
-  set(CMAKE_BUILD_TYPE Debug CACHE STRING
-      "Choose the type of build: None Debug Release RelWithDebInfo MinSizeRel."
-       FORCE)
-endif (NOT CMAKE_BUILD_TYPE)
-
 set(CMAKE_CXX_STANDARD 17)
-set(CMAKE_CXX_FLAGS_DEBUG "${CMAKE_CXX_FLAGS_DEBUG} -std=c++17 -Wall -Wno-unknown-pragmas")
-set(CMAKE_CXX_FLAGS_RELEASE "${CMAKE_CXX_FLAGS_RELEASE} -std=c++17 -O3 -s -Wno-unknown-pragmas")
+set(CMAKE_CXX_STANDARD_REQUIRED ON)
+set(CMAKE_CXX_EXTENSIONS OFF)
 
 set(proc_image_processing_SRC_DIR "src")
-set(proc_image_processing_IMPL_DIR "src/proc_image_processing")
+set(proc_image_processing_CPU_SRC_DIR "src/proc_image_processing/cpu")
+set(proc_image_processing_GPU_SRC_DIR "src/proc_image_processing/gpu")
+
+function(cpu_mode)
+    find_package(OpenCV REQUIRED)
+    catkin_package(
+            INCLUDE_DIRS ${proc_image_processing_SRC_DIR}
+            LIBRARIES
+            CATKIN_DEPENDS
+            roscpp
+            std_msgs
+            image_transport
+            cv_bridge
+            roslaunch
+            pcl_ros
+            sonia_common
+    )
+
+    file(GLOB_RECURSE proc_image_processing_CPU_FILES "${proc_image_processing_CPU_SRC_DIR}/*.cc" "${proc_image_processing_CPU_SRC_DIR}/*.h")
+
+    list(REMOVE_ITEM proc_image_processing_CPU_FILES ${CMAKE_CURRENT_SOURCE_DIR}/${proc_image_processing_CPU_SRC_DIR}/main.cc)
+
+    include_directories(
+            ${catkin_INCLUDE_DIRS}
+            ${proc_image_processing_SRC_DIR}
+            ${sonia_common_INCLUDE_DIRS}
+            ${pcl_ros_INCLUDE_DIRS}
+            ${OpenCV_INCLUDE_DIRS}
+    )
+
+    add_executable(
+            ${PROJECT_NAME}_node
+            ${proc_image_processing_CPU_SRC_DIR}/main.cc
+            ${proc_image_processing_CPU_FILES}
+    )
+
+    target_link_libraries(
+            ${PROJECT_NAME}_node
+            ${catkin_LIBRARIES}
+            ${sonia_common_LIBRARIES}
+            ${pcl_ros_LIBRARIES}
+            ${OpenCV_LIBRARIES}
+            yaml-cpp
+    )
+
+    add_dependencies(${PROJECT_NAME}_node ${catkin_EXPORTED_TARGETS})
+endfunction()
 
+function(gpu_mode)
+    set(OpenCV_DIR /usr/local)
+    find_package(OpenCV REQUIRED NO_MODULE PATHS /usr/local NO_DEFAULT_PATH)
+    include_directories(${OpenCV_INCLUDE_DIRS})
+    add_subdirectory(tools/)
+    catkin_package(
+            INCLUDE_DIRS ${proc_image_processing_SRC_DIR}
+            LIBRARIES
+            CATKIN_DEPENDS
+            roscpp
+            std_msgs
+            image_transport
+            cv_bridge
+            roslaunch
+            pcl_ros
+            sonia_common
+    )
+
+    file(GLOB_RECURSE proc_image_processing_GPU_FILES "${proc_image_processing_GPU_SRC_DIR}/*.cc" "${proc_image_processing_GPU_SRC_DIR}/*.h")
+    file(GLOB_RECURSE proc_image_processing_CPU_FILES "${proc_image_processing_CPU_SRC_DIR}/*.cc" "${proc_image_processing_CPU_SRC_DIR}/*.h")
```

Comment:
I'm not sure why the cpu src files should be included in the gpu_mode function ?

Reply:
Is it because of a fallback mechanism on cpu mode when no compatible gpu is detected ?

Reply:
GPU mode can use stuff from CPU mode, but not the other way around